src/org/jgroups/raft/blocks/LockService.java

```diff
@@ -0,0 +1,808 @@
+		@Override
+		public boolean tryLock(long timeout, TimeUnit unit) throws InterruptedException {
+			long deadline = System.nanoTime() + unit.toNanos(timeout), ns;
+			if (!delegate.tryLock(timeout, unit)) return false;
+			acquirers.incrementAndGet();
+			while (status != HOLDING && (ns = deadline - System.nanoTime()) > 0) {
+				try {
+					if (status == WAITING) notWaiting.awaitNanos(ns);
+					else status = join(LockService.this.lock(lockId));
+				} catch (InterruptedException e) {
+					throw unlock(e);
+				} catch (Throwable e) {
+					rethrow(unlock(e));
+				}
+			}
+			if (status == HOLDING) {
+				holder = Thread.currentThread(); return true;
+			}
+			unlock(); return false;
+		}
+
+		/**
+		 * @throws RaftException if exception happens during sending or executing commands in the lock service.
+		 */
+		@Override
+		public void unlock() {
+			if (!delegate.isHeldByCurrentThread()) return;
+			assert holder == null || holder == Thread.currentThread();
+			if (delegate.getHoldCount() == 1) holder = null;
+			try {
+				if (acquirers.decrementAndGet() == 0 && status != NONE) {
+					join(LockService.this.unlock(lockId));
+					status = NONE;
+				}
+			} catch (Throwable e) {
+				rethrow(e);
+			} finally {
+				delegate.unlock();
+			}
+		}
+
+		/**
+		 * Unsupported
+		 */
+		@Override
+		public Condition newCondition() {
+			throw new UnsupportedOperationException();
+		}
+
+		private <T extends Throwable> T unlock(T error) {
+			try {
+				unlock();
+			} catch (Throwable e) {
+				error.addSuppressed(e);
+			}
+			return error;
+		}
+
+		private <T> T join(CompletableFuture<T> future) throws ExecutionException, TimeoutException {
+			long nanos = TimeUnit.MILLISECONDS.toNanos(timeout), deadline = System.nanoTime() + nanos;
+			boolean interrupted = Thread.interrupted();
+			try {
+				do {
+					try {
+						return future.get(nanos, TimeUnit.NANOSECONDS);
+					} catch (InterruptedException e) {
+						interrupted = true;
+					}
+				} while ((nanos = deadline - System.nanoTime()) > 0);
+				throw new TimeoutException();
+			} finally {
+				if (interrupted) Thread.currentThread().interrupt();
+			}
+		}
+
+		void onStatusChange(long lockId, LockStatus prev, LockStatus curr) {
+			if (lockId != this.lockId) return;
+			if (curr != HOLDING && holder != null) {
+				status = curr;
+				var handler = unlockHandler;
+				if (handler != null) try {
+					handler.accept(this);
+				} catch (Throwable e) {
+					log.error("Error occurred on unlock handler", e);
+				}
+			} else if (curr != NONE && acquirers.get() == 0) {
+				status = curr;
+				var handler = lockHandler;
+				if (handler != null) try {
+					handler.accept(this);
+				} catch (Throwable e) {
+					log.error("Error occurred on lock handler", e);
+				}
+			} else if (prev == WAITING) {
+				delegate.lock();
+				try {
+					if (status == WAITING) {
+						status = curr;
+						notWaiting.signalAll();
+					}
+				} finally {
+					delegate.unlock();
+				}
+			}
```

Comment:
The notified status could be far behind returned status from command executing, it can't be directly set. But WAITTING status is special, if mutex is in WAITTING status, then all calling threads will wait the notification, before that no further commands will be executed.

Reply:
Mutex is just a `Lock` representation of lock service. I think from user's point of view, his requirement is either exclusive thread or exclusive process, even if both are used, they can be separated by different lockIds.

Reply:
It makes sense to make this update only via notifications from the LockService when handling commands. This might still be tricky when installing/recovering from the persistent state. The delegate lock might not be held, which could lead to some internal inconsistency. The mutex must be completely deterministic, regardless of whether or not commands are handled from the log or the user. I believe that making the updates come from the LockService might help.

Reply:
I have changed the notifications source from log applying to command response, log applying is just a tip for the client to query the latest state from server. For now, the local state and notifications only base on command response, pure client mode, seems not bad, please have a look.